src/math/log_e.c

```diff
@@ -0,0 +1,194 @@
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v = numerator << shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts = shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v <<= shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v <<= shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	/*  The input has been left-shifted so the most-significant-bit is a 1. */
+	/*  Reinterpret the output as unsigned with one integer bit, so */
+	/*  that 1 <= x < 2. */
+	x = v << shift_factor;
+	/*  Let Q = int(u).  Then u = Q*2^(-u_fraction_length),	 */
+	/*  and x = Q*2^num_left_shifts * 2^(1-word_length). Therefore, */
+	/*  u = x*2^n, where n is defined as: */
+	/*  Extract the high byte of x */
+	/*  Convert the high byte into an index for lookup table */
+	slice_temp = (v << shift_factor) >> 24;
+	/*  Interpolate between points. */
+	/*  The upper byte was used for the index into lookup table */
+	/*  The remaining bits make up the fraction between points. */
+	a_i = BASE_E_LOG_DENOM_WORDLENGTH - num_left_shifts;
+	a_i <<= 16;
+	l1_i = iv1[(BASE_E_LOG_WRAP_SCHAR_BITS & ((x >> 24) + 129)) - 1];
+	s_v = a_i + l1_i;
+	l2_i = iv1[(BASE_E_LOG_WRAP_SCHAR_BITS & (slice_temp + 130)) - 1];
+	l3_i = iv1[(slice_temp - 127) - 1];
+	a = s_v + (((x & BASE_E_LOG_UPPERBYTES) * (int64_t)(l2_i - l3_i)) >> 24);
+
+	/* calculate denominator-const exp(1)*/
+	shift_factor = iv[denominator >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v = denominator << shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts = shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v <<= shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	v <<= shift_factor;
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	shift_factor = iv[v >> 24];
+	/*  Left-shift out all the leading zeros in the high byte. */
+	/*  Update the total number of left-shifts */
+	num_left_shifts += shift_factor;
+	/*  The input has been left-shifted so the most-significant-bit is a 1. */
+	/*  Reinterpret the output as unsigned with one integer bit, so */
+	/*  that 1 <= x < 2. */
+	x = v << shift_factor;
+	slice_temp = (v << shift_factor) >> 24;
+	a_i = BASE_E_LOG_NUMER_WORDLENGTH - num_left_shifts;
+	a_i <<= 16;
+	l1_i = iv1[(BASE_E_LOG_WRAP_SCHAR_BITS & ((x >> 24) + 129)) - 1];
```

Comment:
Can we use the ARRAY_SIZE() here instead of 129 (and in other places). Btw, what checking do we do for valid array index ?

Reply:
how close is this code to log_10? It looks very similar. If its the same we should just create a helper, pass in the table and iniline the main call

Reply:
The comment mentions both are based on log2(x) function. Then I'd expect the log10() and log() functions to use a common log2(x) function and divide the result with log2(10) or log2(e) constants.

Reply:
Done. removed code duplication , now it's one liner.